File: CS/1379_Find_a_Corresponding_Node_of_a_Binary_Tree_in_a_Clone_of_That_Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
    public TreeNode GetTargetCopy(TreeNode original, TreeNode cloned, TreeNode target) {
        //TreeNode result = new TreeNode();
        if(original == null){
            return null;
        }
        else if(original == target){
            return cloned;
        }
        
        TreeNode left = GetTargetCopy(original.left, cloned.left, target);
        TreeNode right = GetTargetCopy(original.right, cloned.right, target);
        return left == null? right: left;
    }
}

File: Cpp/1379_Find_a_Corresponding_Node_of_a_Binary_Tree_in_a_Clone_of_That_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* result;
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        if(!original){
            return nullptr;
        }
        
        if(original == target){
            result = cloned;
        }
        else{
            getTargetCopy(original->left, cloned->left, target);
            getTargetCopy(original->right, cloned->right, target);
            /*
            TreeNode* left = getTargetCopy(original->left, cloned->left, target);
            return left != nullptr ? left : getTargetCopy(original->right, cloned->right, target);
            */
        }
        return result;
    }
};
